packages/clarity-js/src/layout/region.ts

```diff
@@ -1,108 +1,77 @@
-import { Event, BooleanFlag } from "@clarity-types/data";
-import { Box, RegionData } from "@clarity-types/layout";
+import { BooleanFlag, Event } from "@clarity-types/data";
+import { RegionData } from "@clarity-types/layout";
+import * as dom from "@src/layout/dom";
 import encode from "@src/layout/encode";
-import * as dom from "./dom";
 
-let bm: { [key: number]: RegionData } = {};
-let updateMap: number[] = [];
+export let data: RegionData[] = [];
+let regionMap: WeakMap<Node, string> = null; // Maps region nodes => region name
+let regions: { [key: number]: RegionData } = {};
+let updates: number[] = [];
+let watch = false;
+let observer: IntersectionObserver = null;
 
-export function compute(): void {
-    // Making region compute synchronous for tracking regions 
-    // TODO: For modern browsers use resizeobserver 
-    region();
-}
-
-async function region(): Promise<void> {
-    let values = dom.regions();
-    let viewport = getViewport();
-
-    for (let value of values) {
-        update(value.id, layout(dom.getNode(value.id) as Element, viewport));
-    }
-
-    if (updateMap.length > 0) { await encode(Event.Region); }
+export function start(): void {
+    reset();
+    observer = null;
+    regionMap = new WeakMap();
+    regions = {};
+    watch = window["IntersectionObserver"] ? true : false;
+    
 }
 
-export function updates(): RegionData[] {
-    let summary = [];
-    for (let id of updateMap) {
-        summary.push(bm[id]);
+export function observe(node: Node, name: string): void {
+    if (regionMap.has(node) === false) {
+        regionMap.set(node, name);
```

Comment:
This will use the default threshold of zero for the thresholds. Will that causes a lot of events and would a higher threshold be appropriate? By "lot of events" I mean on the client and also events we upload.

Reply:
Zero is intentional because it's hard to come up with a magic number otherwise. We tweak this later but this worked best when dealing with regions of various sizes. For instance 10% of "body" may not trigger visibility if the page was long and that will be inaccurate to say that body wasn't visible. Also, it won't result in more events. If threshold is zero and once it becomes visible, next event only fires when it gets hidden. Basically no chatter as it goes from 0% -> 10% -> 50% -> 100%.